addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+
+    /** Plugwise protocol trailer code (hex) */
+    private static final String PROTOCOL_TRAILER = new String(new char[] { CR, LF });
+
+    /** Matches Plugwise responses into the following groups: protocolHeader command sequence payload CRC */
+    private static final Pattern RESPONSE_PATTERN = Pattern.compile("(.{4})(\\w{4})(\\w{4})(\\w*?)(\\w{4})");
+
+    /** Default maximum number of attempts to send a message */
+    private static final int MAX_RETRIES = 1;
+
+    /** After exceeding this threshold the Stick is set offline */
+    private static final int MAX_SEQUENTIAL_WRITE_ERRORS = 15;
+
+    // Queue fields
+    private static final int MAX_BUFFER_SIZE = 1024;
+
+    private final Logger logger = LoggerFactory.getLogger(PlugwiseCommunicationHandler.class);
+    private final MessageFactory messageFactory = new MessageFactory();
+    private final List<FilteredMessageListener> filteredListeners = new CopyOnWriteArrayList<>();
+
+    private PlugwiseStickConfig configuration;
+
+    private final PriorityBlockingQueue<QueuedMessage> sendQueue = new PriorityBlockingQueue<>(MAX_BUFFER_SIZE,
+            QUEUED_MESSAGE_COMPERATOR);
+    private final BlockingQueue<AcknowledgementMessage> acknowledgedQueue = new ArrayBlockingQueue<>(MAX_BUFFER_SIZE,
+            true);
+    private final BlockingQueue<QueuedMessage> sentQueue = new ArrayBlockingQueue<>(MAX_BUFFER_SIZE, true);
+    private final BlockingQueue<Message> receivedQueue = new ArrayBlockingQueue<>(MAX_BUFFER_SIZE, true);
+
+    private final ReentrantLock sentQueueLock = new ReentrantLock();
+
+    // Background threads
+    private Thread messageProcessor;
+    private Thread messageSender;
+
+    // Serial communication fields
+    private ByteBuffer readBuffer = ByteBuffer.allocate(MAX_BUFFER_SIZE);
+    private SerialPort serialPort;
+    private WritableByteChannel outputChannel;
+    private int previousByte = -1;
+    private int sequentialWriteErrors = 0;
+
+    private boolean initialized = false;
+
+    public void addMessageListener(PlugwiseMessageListener listener) {
+        if (!isExistingMessageListener(listener)) {
+            filteredListeners.add(new FilteredMessageListener(listener));
+        }
+    }
+
+    public void addMessageListener(PlugwiseMessageListener listener, MACAddress macAddress) {
+        if (!isExistingMessageListener(listener, macAddress)) {
+            filteredListeners.add(new FilteredMessageListener(listener, macAddress));
+        }
+    }
+
+    /**
+     * Initialize this device and open the serial port
+     *
+     * @throws PlugwiseInitializationException if port can not be opened
+     */
+    @SuppressWarnings("rawtypes")
+    private void initializeSerialPort() throws PlugwiseInitializationException {
+
+        CommPortIdentifier portId = null;
+
+        // Parse ports and if the default port is found, initialized the reader
+        Enumeration portList = CommPortIdentifier.getPortIdentifiers();
```

Comment:
Did you copy this from one of the other bindings, should you credit people for that?